File: solutions/two-pointers/011-container-with-most-water.py
"""
# 011. Container With Most Water
**Medium**

Given a problem that demonstrates key concepts in Two Pointers.

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of two pointers concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply two pointers methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages two pointers principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### TIME COMPLEXITY: O(n)
### SPACE COMPLEXITY: O(1)

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>

<details>
<summary><b>💡 APPROACH</b></summary>

The approach uses two pointers techniques to solve this problem efficiently.

### Algorithm Steps:
1. Initialize necessary variables
2. Process input using two pointers method
3. Return the computed result

</details>
"""

class Solution:
    def maxArea(self, height: List[int]) -> int:
        """
        Calculate the maximum area of water that can be contained between two vertical lines.
        
        Args:
            height: List of integers representing heights of vertical lines
            
        Returns:
            Maximum area of water that can be contained
        """
        # Initialize pointers and maximum area
        left = 0
        right = len(height) - 1
        max_area = 0
        
        # Use two-pointer technique to find maximum area
        while left < right:
            # Calculate current area
            # Area = width * minimum height between two lines
            current_area = (right - left) * min(height[left], height[right])
            
            # Update maximum area if current area is larger
            max_area = max(max_area, current_area)
            
            # Move the pointer with smaller height inward
            # This is optimal because moving the larger height pointer
            # would only decrease the area
            if height[left] < height[right]:
                left += 1
            else:
                right -= 1
                
        return max_area

def test_solution():
    """
    Test cases for 011. Container With Most Water.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 011. Container With Most Water")

File: solutions/two-pointers/alternatives/011-container-with-most-water.js
/**
 * Difficulty: Medium
 *
 * [Problem description goes here]
 *
 * **Example:**
 *
 * <dl class="example-details">
 * <dt>Input:</dt>
 * <dd>[input description]</dd>
 * <dt>Output:</dt>
 * <dd>[output description]</dd>
 * <dt>Explanation:</dt>
 * <dd>[explanation]</dd>
 * </dl>
 *
 * <details>
 * <summary><b>🔍 SOLUTION EXPLANATION</b></summary>
 *
 * ### INTUITION:
 * [High-level insight or key observation]
 *
 * ### APPROACH:
 * [Detailed explanation of the solution approach]
 *
 * ### WHY THIS WORKS:
 * - [Explanation of correctness]
 *
 * ### EXAMPLE WALKTHROUGH:
 * Input:
 * ```
 * [example input]
 * ```
 *
 * **Step 1:** [description]
 *
 * **Step 2:** [description]
 *
 * ### TIME COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### SPACE COMPLEXITY:
 * **O(?)** - [explanation]
 *
 * ### EDGE CASES:
 * - **[Edge case 1]:** [how it's handled]
 * - **[Edge case 2]:** [how it's handled]
 *
 * </details>
 */

/**
 * Main solution for Problem 011: Container With Most Water
 *
 * @param {number[]} height - Array of non-negative integers representing heights
 * @return {number} - Maximum area of water that can be contained
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
function solve(height) {
    // Edge case: need at least 2 lines to contain water
    if (!height || height.length < 2) {
        return 0;
    }

    let left = 0;
    let right = height.length - 1;
    let maxArea = 0;

    while (left < right) {
        // Calculate current area
        const width = right - left;
        const currentHeight = Math.min(height[left], height[right]);
        const currentArea = width * currentHeight;

        // Update maximum area
        maxArea = Math.max(maxArea, currentArea);

        // Move the pointer pointing to the shorter line
        if (height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }

    return maxArea;
}

/**
 * Test cases for Problem 011: Container With Most Water
 */
function testSolution() {
    console.log('Testing 011. Container With Most Water');

    // Test case 1: Example from problem
    const result1 = solve([1, 8, 6, 2, 5, 4, 8, 3, 7]);
    const expected1 = 49;
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);
    console.log(`Test 1 passed: maxArea([1,8,6,2,5,4,8,3,7]) = ${result1}`);

    // Test case 2: Two lines only
    const result2 = solve([1, 1]);
    const expected2 = 1;
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);
    console.log(`Test 2 passed: maxArea([1,1]) = ${result2}`);

    // Test case 3: Increasing heights
    const result3 = solve([1, 2, 3, 4, 5]);
    const expected3 = 6;
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);
    console.log(`Test 3 passed: maxArea([1,2,3,4,5]) = ${result3}`);

    // Test case 4: Decreasing heights
    const result4 = solve([5, 4, 3, 2, 1]);
    const expected4 = 6;
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);
    console.log(`Test 4 passed: maxArea([5,4,3,2,1]) = ${result4}`);

    // Test case 5: Edge case - empty array
    const result5 = solve([]);
    const expected5 = 0;
    console.assert(result5 === expected5, `Test 5 failed: expected ${expected5}, got ${result5}`);
    console.log(`Test 5 passed: maxArea([]) = ${result5}`);

    // Test case 6: Edge case - single element
    const result6 = solve([1]);
    const expected6 = 0;
    console.assert(result6 === expected6, `Test 6 failed: expected ${expected6}, got ${result6}`);
    console.log(`Test 6 passed: maxArea([1]) = ${result6}`);

    console.log('All test cases passed for 011. Container With Most Water!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 011. Container With Most Water ===');
    console.log('Category: Two Pointers');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on two pointers concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */
